### PR TITLE
Expose StoredKey encryption parameters

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Expose StoredKey encryption parameters, through C interface, as Json string.

StoredKey unit tests.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).

### CODE REVIEW

File: swift/Tests/Keystore/KeystoreKeyTests.swift [hunk 1]
```diff
@@ -119,4 +119,25 @@ class KeystoreKeyTests: XCTestCase {
         let data = keystore.decryptPrivateKey(password: password)
         XCTAssertEqual(data?.hexString, "4357b2f9a6150ba969bc52f01c98cce5313595fe49f2d08303759c73e5c7a46c")
     }
+
+    struct KdfParamsStruct: Decodable {
```
Comment (hewigovens) [inline swift/Tests/Keystore/KeystoreKeyTests.swift:123]:
```suggestion
    struct KdfParams: Decodable {
```

File: swift/Tests/Keystore/KeystoreKeyTests.swift [hunk 2]
```diff
@@ -119,4 +119,25 @@ class KeystoreKeyTests: XCTestCase {
...
+        let dklen: Int
+        let n: Int
+    }
+
+    struct EncryptionParametersStruct: Decodable {
```
Comment (hewigovens) [inline swift/Tests/Keystore/KeystoreKeyTests.swift:128]:
```suggestion
    struct EncryptionParameters: Decodable {
```

File: swift/Tests/Keystore/KeystoreKeyTests.swift [hunk 3]
```diff
@@ -119,4 +119,25 @@ class KeystoreKeyTests: XCTestCase {
...
+    func testEncryptionParameters() {
+        let url = Bundle(for: type(of: self)).url(forResource: "key", withExtension: "json")!
+        let key = StoredKey.load(path: url.path)!
+
+        let params = key.encryptionParameters
+        let paramsData = params!.data(using: .utf8)!
+        let paramsSruct: EncryptionParametersStruct = try! JSONDecoder().decode(EncryptionParametersStruct.self, from: paramsData)
+        XCTAssertEqual(paramsSruct.kdf, "scrypt");
+        XCTAssertEqual(paramsSruct.kdfparams.n, 262144);
+    }
```
Comment (hewigovens) [inline swift/Tests/Keystore/KeystoreKeyTests.swift:142]:
```suggestion
    func testEncryptionParameters() throws {
        let url = Bundle(for: type(of: self)).url(forResource: "key", withExtension: "json")!
        let key = StoredKey.load(path: url.path)!

        let paramsData = key.encryptionParameters!.data(using: .utf8)!
        let params = try JSONDecoder().decode(EncryptionParameters.self, from: paramsData)
        
        XCTAssertEqual(params.kdf, "scrypt");
        XCTAssertEqual(params.kdfparams.n, 262144);
    }
```

File: tests/interface/TWStoredKeyTests.cpp [hunk 4]
```diff
@@ -225,3 +226,36 @@ TEST(TWStoredKey, getWalletPasswordInvalid) {
...
+TEST(TWStoredKey, encryptionParameters) {
+    const auto key = createDefaultStoredKey();
+    const auto params = WRAPS(TWStoredKeyEncryptionParameters(key.get()));
+
+    nlohmann::json jsonParams = nlohmann::json::parse(string(TWStringUTF8Bytes(params.get())));
+
+    // compare one parameter
+    EXPECT_EQ(jsonParams["kdfparams"]["n"], 4096);
+
+    // compare all keys, except dynamic ones (like cipherparams/iv)
```
Comment (hewigovens) [inline tests/interface/TWStoredKeyTests.cpp:239]: for iv/cipherparams, we can test length is large than some value

File: swift/Tests/Keystore/KeystoreKeyTests.swift (hunk 1 above)
Comment (catenocrypt) [reply swift/Tests/Keystore/KeystoreKeyTests.swift]: thx, done

File: swift/Tests/Keystore/KeystoreKeyTests.swift (hunk 2 above)
Comment (catenocrypt) [reply swift/Tests/Keystore/KeystoreKeyTests.swift]: thx, done

File: swift/Tests/Keystore/KeystoreKeyTests.swift [hunk 5]
```diff
@@ -119,4 +119,25 @@ class KeystoreKeyTests: XCTestCase {
...
+        let dklen: Int
+        let n: Int
+    }
+
+    struct EncryptionParameters: Decodable {
+        let kdf: String
+        let kdfparams: KdfParams
+    }
+
+    func testEncryptionParameters() {
```
Comment (vikmeup) [inline swift/Tests/Keystore/KeystoreKeyTests.swift:133]: can you show an example how to encrypt with specific params?

**Review by hewigovens — APPROVED** [review]

File: swift/Tests/Keystore/KeystoreKeyTests.swift (hunk 5 above)
Comment (catenocrypt) [reply swift/Tests/Keystore/KeystoreKeyTests.swift:133]: I have checked into this, and currently this is not possible.  Always default encryption parameters are used, and there is no way to influence that.
If this is needed, let's clarify and create a separate issue.

File: swift/Tests/Keystore/KeystoreKeyTests.swift [hunk 6]
```diff
@@ -119,4 +119,25 @@ class KeystoreKeyTests: XCTestCase {
         let data = keystore.decryptPrivateKey(password: password)
         XCTAssertEqual(data?.hexString, "4357b2f9a6150ba969bc52f01c98cce5313595fe49f2d08303759c73e5c7a46c")
     }
+
+    struct KdfParams: Decodable {
+        let dklen: Int
+        let n: Int
+    }
+
+    struct EncryptionParameters: Decodable {
```
Comment (vikmeup) [inline swift/Tests/Keystore/KeystoreKeyTests.swift:128]: can we have ability to specify `EncryptionParameters` on wallet creation?

File: swift/Tests/Keystore/KeystoreKeyTests.swift [hunk 7]
```diff
@@ -119,4 +119,25 @@ class KeystoreKeyTests: XCTestCase {
...
+        let kdf: String
+        let kdfparams: KdfParams
```
Comment (vikmeup) [inline swift/Tests/Keystore/KeystoreKeyTests.swift:130]: can we specify `StrongEncryptionParameters` type? and default

**Comment by catenocrypt** [comment]
Encryption parameters can be controlled from outside:
- There is an enum with values Default, Weak and Standard   https://github.com/trustwallet/wallet-core/pull/1766/files#diff-ff71e2e6bd79b5044dd5c8c11a0ae08cf4df0413d51ecb8f9b4d074bcff5e590
- When creating StoredKey wallet storage, level can be specified
- Weak level is with 4096, used currently
- Standard level is with 262k.

File: swift/Tests/Keystore/KeystoreKeyTests.swift (hunk 6 above)
Comment (catenocrypt) [reply swift/Tests/Keystore/KeystoreKeyTests.swift:128]: Done, see below

**Comment by vikmeup** [comment]
> Weak level is with 4096, used currently

let's change it to 10K as it's usually a default option in many places

**Review by vikmeup — APPROVED** [review]

**Review by hewigovens — APPROVED** [review]

**Comment by catenocrypt** [comment]
> > Weak level is with 4096, used currently
> 
> let's change it to 10K as it's usually a default option in many places

I have added a new level, with 16k (power of 2); #1801 .

**Pull request merged**